data sampling refactor + online eval

katerakelly opened the pull request:
submitting as one PR because disentangling some of the commits was going to require a lot of re-writing

katerakelly committed:
implement online update for information bottleneck case

update by summing natural parameter
TODO: switch to predicting natural parameter

katerakelly committed:
simplify and generalize data sampling

 - parameterize data sampling by number of transitions, rate to
resample z, rate to update posterior q(z|c), batch_size to use for
update
 - in the model, decouple posterior update from sampling z

set default to recompute posterior and resample z every trajectory
 of data collected
add args to control this

katerakelly committed:
generalize evaluation to do online posterior inference

always update the running latent context on every transition
resample z at different intervals (e.g. transition-level, traj-level
adaptation)

remove the eval encoder replay buffer as it's no longer needed

katerakelly committed:
use stochastic policy during eval to match training

katerakelly committed:
add eval of train tasks with train replay buffer

helps discern between optimization and generalization issues

katerakelly committed:
seed generation of random point robot goals

katerakelly committed:
sample sequences with replacement, just like transitions

rename trajs -> sequence

azhou42 commented on rlkit/envs/point_robot.py:
```diff
@@ -8,13 +8,17 @@
     def __init__(self, randomize_tasks=False, n_tasks=2):
 
         if randomize_tasks:
+            np.random.seed(1337)
```
but why

katerakelly replied:
my old collaborator used it for everything and I picked up the habit haha XD

azhou42 commented on rlkit/torch/sac/proto.py:
```diff
@@ -16,26 +16,42 @@ def _product_of_gaussians(mus, sigmas_squared):
                  **kwargs
     ):
         super().__init__()
+        self.env = env
```
I don't feel too strongly, but I don't think the agent should have access to the env here. If all we're doing is calling sparsify reward, it's probably cleaner to just store it as an extra field in the replay buffer. It also just seems pretty hacky to me that the sparse reward is just a function of the original reward.

katerakelly replied:
I agree - we can do what I did when I was trying to make the MAESN envs work - stuff the sparse reward into `env_infos` (adding another field to the replay buffer is a pain).

katerakelly replied:
see `9aaf099`

azhou42 commented on rlkit/torch/sac/proto.py:
```diff
@@ -54,90 +70,116 @@ def __init__(self,
+            # only have one task here
+            up_params = up_params[0] # batch x feat
+            curr_mu = self.z_means[0]
+            curr_ss = self.z_vars[0]
```
will z_vars[0] exist if we're not using ib? this might crash

katerakelly replied:
yes, in the deterministic case, `z_vars` is set to zero (see line 77)

azhou42 commented on rlkit/torch/sac/proto.py:
```diff
@@ -54,90 +70,116 @@ def __init__(self,
+            else:
+                for i in range(num_updates):
+                    num_z += 1
+                    curr_mu += (up_params[i] - curr_mu) / num_z
```
this computation seems bugged. if it's actually the mean, then it should be 
(x + (n-1) * mu) / n?

katerakelly replied:
it's equivalent. written this way it's just factorized into the sum of the old value and the new part

azhou42 commented on rlkit/torch/torch_rl_algorithm.py:
```diff
@@ -46,187 +46,130 @@ def get_batch(self, idx=None):
-            train_avg_returns.append(eval_util.get_average_returns(paths))
+        ### sample trajectories from prior for vis
+        prior_paths = []
+        for _ in range(10):
```
make 10 a parameter?

katerakelly replied:
or better yet I think this should go into `debug` mode since aside from visualization it's just a waste of computation

katerakelly replied:
see `258b316`...it's hard-coded to collecting 100 trajs, but that seems ok to me

azhou42 replied:
ya seems fine if its just for debugging

katerakelly committed:
fix stray deprecated argument

katerakelly committed:
collect samples from prior only when dumping eval paths

only useful for debugging or visualization

azhou42 commented on examples/default.py:
```diff
@@ -36,10 +36,11 @@ def make_variant(max_path_length):
-            # {'initial_pool', 'online_exploration_trajectories', 'online_on_policy_trajectories'}
-            eval_embedding_source='online_exploration_trajectories',
+            resample_z='trajectory', # how often to resample z at test-time
+            resample_z_train=max_path_length, # how often to resample z at training time
```
Seems weird to have resample_z take value 'trajectory', and resample_z_train=max path length.
I think we should maintain a consistent structure for controlling how embeddings are done at train and test.

katerakelly replied:
I agree...my original plan was to have all resample at any arbitrary rate. I'm not sure how to handle episodes that terminate before `max_path_length` though. Should we have a special argument value that means "trajectory" however long that happens to be? Or we could go the other way and have just transitions and multiples of trajectories

azhou42 replied:
We could just have it consistently resample by trajectory instead of max_path_length?

katerakelly replied:
done in `de1b5c3`

azhou42 commented on rlkit/core/rl_algorithm.py:
```diff
@@ -189,31 +185,17 @@ def train(self):
-                elif self.train_embedding_source == 'posterior_only':
-                    self.collect_data_from_task_posterior(idx=idx, num_samples=self.num_steps_per_task, eval_task=False,
-                                                          add_to_enc_buffer=True)
+                    self.collect_data(self.num_steps_per_task, self.resample_z_train, np.inf, self.embedding_batch_size, add_to_enc_buffer=False)
```
I think it's also okay to just get rid of this option altogether. I don't see when someone would use it at this point.

katerakelly replied:
do you think the same is true of `initial_pool`?

azhou42 replied:
This comment was addressed towards the initial_pool option.

katerakelly replied:
see `9161b6b`

azhou42 commented on rlkit/core/rl_algorithm.py:
```diff
@@ -224,7 +206,7 @@ def train(self):
             gt.stamp('train')
 
-            #self.training_mode(False)
+            self.training_mode(False)
```
Do we actually use this flag now?

katerakelly replied:
ah I meant to make this a separate commit to highlight it....it breaks the LSTM for reasons unknown to me but I think it's unsafe to omit it unless we have to, and I don't think `master` needs to support the LSTM

azhou42 commented on rlkit/core/rl_algorithm.py:
```diff
@@ -238,74 +220,31 @@ def pretrain(self):
+        for _ in range(num_samples // resample_z_rate):
+            self.collect_transitions(self.policy, num_samples=resample_z_rate, add_to_enc_buffer=add_to_enc_buffer)
+            counter += resample_z_rate
+            if counter % update_posterior_rate == 0:
```
If for some reason resample_z_rate does not divide update_posterior_rate, this doesn't work as it is probably intended. Also, given that we reset the envs between switching tasks, maybe collect transitions should just collect complete trajectories until we exceed num_samples (returning the # of smaples it consumed), and then update posterior if stepcount >= update_count * posterior rate.

katerakelly replied:
see `81f27a8`

azhou42 commented on rlkit/samplers/util.py:
```diff
@@ -51,6 +52,8 @@ def rollout(env, agent, max_path_length=np.inf, animated=False, is_online=False)
         o = next_o
         if animated:
             env.render()
+    if resample == 'trajectory':
```
personally don't think this should belong in the rollout method (should just be called in the while loop in obtain_samples)

katerakelly replied:
I disagree because I think all the resampling should happen in the same function, and transition-level resampling has to happen here. Though if we generalize it to resample after any number of transitions, we'll run into this problem again...

azhou42 replied:
My reasoning on keeping it out is that rollout shouldn't care about what happens after the rollout, but I'm fine either way. I'm against resampling after arbitrary numbers of transitions. I think it should just be sample whole trajectories until we reach a threshold to resample.

katerakelly replied:
ok I think we should just sample by trajectory at test time to keep things consistent, so I'll move this out of the rollout function

katerakelly replied:
see `da9217c`

katerakelly committed:
remove initial_pool option from data sampling

katerakelly committed:
sample whole trajectories during training data collection

this way episodes always terminate properly

azhou42 commented on rlkit/core/rl_algorithm.py:
```diff
@@ -217,41 +217,39 @@ def pretrain(self):
-            self.collect_transitions(self.policy, num_samples=resample_z_rate, add_to_enc_buffer=add_to_enc_buffer)
-            counter += resample_z_rate
-            if counter % update_posterior_rate == 0:
+            num_collected += self.collect_transitions(self.policy, num_samples=resample_z_rate, add_to_enc_buffer=add_to_enc_buffer)
```
collect_transitions should be renamed to collect_trajectories

katerakelly replied:
see `de1b5c3`

azhou42 commented on rlkit/core/rl_algorithm.py:
```diff
@@ -217,41 +217,39 @@ def pretrain(self):
             if self.render:
                 self.env.render()
-            next_ob, raw_reward, terminal, env_info = (
+            next_ob, reward, terminal, env_info = (
```
do you know why it was originally raw reward here?

katerakelly replied:
yes because reward scaling used to be done here -_-

azhou42 replied:
😱

azhou42 commented on rlkit/core/rl_algorithm.py:
```diff
@@ -265,14 +263,16 @@ def collect_transitions(self, agent, num_samples=1, add_to_enc_buffer=True):
             )
-            if terminal or len(self._current_path_builder) >= self.max_path_length:
+            num_collected += 1
+            if terminal:
```
are you sure we don't need to check the max path length on the algorithm side?

katerakelly replied:
you're right, I accidentally deleted that! fixed in `de1b5c3`

katerakelly committed:
change training data sampling to update posterior/sample z by traj

it's more realistic to resample z and recompute the posterior between
trajectories rather than arbitrarily

katerakelly committed:
resample z at test-time at the trajectory level

this is aligned with training data collection

katerakelly committed:
store sparse rewards in env_infos

much simpler to use sparse rewards for the encoder

katerakelly committed:
fix: use mod operator correctly

katerakelly committed:
err on the side of over-sampling when collecting eval trajs

katerakelly commented on rlkit/core/rl_algorithm.py:
```diff
@@ -175,45 +174,23 @@ def train(self):
-                                                          resample_z_every_n=self.max_path_length,
-                                                          eval_task=False)
-
+                    self.collect_data(self.policy, self.num_initial_steps, 1, np.inf)
```
I kind of dislike using `np.inf` to get samples from the prior only, but not sure what the cleaner way should look like?

azhou42 replied:
I t hink np inf is fine.

katerakelly replied:
let's roll with it for now

katerakelly commented:
remaining issues: online eval does not give the same results as `master` branch for cheetah-vel

katerakelly committed:
update point robot and cheetah scripts to match how scripts now work on master

katerakelly committed:
FIX: deterministic context should take running mean correctly

katerakelly commented:
final status: online latent context computation doesn't quite match with batch computation - probably a numerical issue - so we will close this PR as is and revert back to encoding context batch-wise in a future PR.
